### DISCUSSION TITLE
Implicit use of "using" when creating a IDisposable object

### DISCUSSION THREAD

**Discussion by Lavshyak** [dotnet/csharplang#6869] [Language Ideas] [answered]
I propose to make the word "using" by default and add the word "notusing" like.

- ### Now:

  When you need to release unmanaged resources when going out of scope
  ```
  void Foo()
  {
  using var someObj = new IDisposableClass();
  someObj.DoSomething();
  }//here is the call to the Dispose method of the someObj object
  ```
  
  When you do **not** need to release unmanaged resources when going out of scope
  ```
  void Foo()
  {
  var someObj = new IDisposableClass();
  someObj.DoSomething();
  }
  ```

- ### After:

  - **ImplicitUsing compillation flag setted**
  
    When you need to release unmanaged resources when going out of scope
    ```
    void Foo()
    {
    var someObj = new IDisposableClass();
    someObj.DoSomething();
    }//here is the call to the Dispose method of the someObj object
    ```

    When you do **not** need to release unmanaged resources when going out of scope
    ```
    void Foo()
    {
    notusing /*some word like that*/ var someObj = new IDisposableClass();
    someObj.DoSomething();
    }
    ```

  
  - **ImplicitUsing compillation flag is not setted**
    (will be the same code as in the "Now" block)

    When you need to release unmanaged resources when going out of scope

    ```
    void Foo()
    {
    using var someObj = new IDisposableClass();
    someObj.DoSomething();
    }//here is the call to the Dispose method of the someObj object
    ```
    
    When you do **not** need to release unmanaged resources when going out of scope
    ```
    void Foo()
    {
    var someObj = new IDisposableClass();
    someObj.DoSomething();
    }
    ```

In most cases, IDisposable class objects are created using the word "using". And this is not very convenient, since you need to write a lot of letters and you have to see if the class implements the IDisposable interface. By making the default word "using", you can get rid of these problems.
With the approach described above, old code will not be affected, but it will be easier to write code in new projects.

**Comment by HaloFour** [comment, marked as answer]
This would result in creating a dialect of the language which would have an impact on runtime behavior, something the team has expressed being opposed to doing.

**Reply by Lavshyak** [reply]
> impact on runtime behavior?
isn't it enough just to add the word "using" to the code or remove word "nousing" at build or compile time?

**Reply by HaloFour** [reply]
> isn't it enough just to add the word "using" to the code or remove word "nousing" at build or compile time?

You can't know what any given code will do without also knowing what the compiler/project settings were.  That's what makes it a dialect, the same code could mean two completely different things.
